symengine/polys/uintpoly.cpp

```diff
@@ -1,7 +1,241 @@
+              std::set<GaloisFieldDict, GaloisFieldDict::DictLess>> fsqf;
+    size_t counter = 0;
+    Sieve::iterator pi(bound);
+    pi.next_prime(); // To avoid 2
+    unsigned i;
+    while ((i = pi.next_prime()) <= bound) {
+        if (b % i == 0)
+            continue;
+        integer_class i_c = integer_class(i);
+        auto F = GaloisFieldDict(this->poly_.dict_, i_c);
+        if (not F.gf_is_sqf())
+            continue;
+        F.gf_monic(outArg(F));
+        auto fsqfx = F.gf_zassenhaus();
+        if (fsqfx.size() < fsqf.second.size() or fsqf.second.empty()) {
+            fsqf.first = i_c;
+            fsqf.second = fsqfx;
+            ++counter;
+        }
+        if (fsqfx.size() < 15 or counter > 4)
+            break;
+    }
+    unsigned l;
+    if (mp_fits_slong_p(B)) {
+        temp = 2 * B + 1;
+        l = std::ceil(std::log(mp_get_d(temp))
+                      / std::log(mp_get_ui(fsqf.first)));
+    } else {
+        auto b_d = mp_get_d(mp_abs(b));
+        l = std::ceil(
+            (1 + std::log2(n + 1) / 2.0 + n + log2_d_A + std::log2(b_d))
+            / (std::log2(mp_get_ui(fsqf.first))));
+    }
+    integer_class pl;
+    mp_pow_ui(pl, fsqf.first, l);
+    GaloisFieldDict gf(this->poly_.dict_, pl);
+    std::vector<UIntDict> g
+        = GaloisFieldDict::zz_hensel_lift(gf, fsqf.first, fsqf.second, l);
+
+    std::vector<unsigned int> T(g.size());
+    std::iota(std::begin(T), std::end(T), 0);
+    unsigned int s = 1;
+
+    auto _test_pl = [](const integer_class &fc, const integer_class &q,
+                       const integer_class &pl) {
+        integer_class q_copy(q);
+        if (q_copy > pl / 2_z)
+            q_copy -= pl;
+        if (q_copy == 0_z)
+            return true;
+        return (fc % q_copy == 0_z);
+    };
+
+    while (2 * s <= T.size()) {
+        bool broken = false;
+        std::string bitmask(s, 1);
+        bitmask.resize(T.size(), 0);
+        do {
+            std::vector<unsigned int> S;
+            for (unsigned int i = 0; i < T.size(); ++i) {
+                if (bitmask[i]) {
+                    S.push_back(T[i]);
+                }
+            }
+            UIntDict G;
+
+            if (b == 1_z) {
+                integer_class q = 1_z;
+                for (auto i : S) {
+                    q *= g[i].dict_.begin()->second;
+                }
+                q %= pl;
+                if (not _test_pl(fc, q, pl))
+                    continue;
+            } else {
+                G = UIntDict(b);
+                for (auto &i : S) {
+                    G *= g[i];
+                }
+                G.itrunc(pl);
+                G.primitive(outArg(G));
+                auto q = G.dict_.begin()->second;
+                if (q != 0 and fc % q != 0)
+                    continue;
+            }
+            UIntDict H = UIntDict(b);
+            std::vector<unsigned int> T_S(T.size());
+            auto it = std::set_difference(T.begin(), T.end(), S.begin(),
```

Comment:
From what I gather, `S` is the set of numbers where bitmask is set to `1` and `T_S` is the set of numbers where bitmask is set to `0` right? If so, why not construct it like that, instead of this complicated way?